content/blog/2023/2023-10-25-spring-java-config.md

```diff
@@ -0,0 +1,1001 @@
+
+### XML Configuration
+It is the classic approach. Configure Spring beans and dependencies using XML files. Perfect for legacy systems and scenarios requiring external configuration.
+
+### Annotation-based Configuration
+It offers simplicity. Use annotations like `@Component`, `@Service`, and `@Repository` to define beans. Ideal for small to medium-sized projects, reducing XML clutter.
+
+### Java-based Configuration
+Configure beans with Java classes using `@Configuration` and `@Bean`. Promotes modularity and testability, making our codebase cleaner. We would focus on this configuration in this article.
+
+### Java Configuration with Annotation Scanning
+Combine Java-based configuration with annotation scanning for effortless bean discovery. Boost productivity in large projects. Manual Java-based configuration can be used in special scenarios where custom configuration and more control is needed. In the sections to follow, we will learn about customizations offered by Java-based configuration.
+
+### Java Configuration with Property Files
+It offers flexibility. Externalize configuration properties to property files. Inject values using `@Value` annotation, ensuring adaptability across environments.
+
+{{% info title="Use Case: Employee Management System" %}}
+Imagine you are building an Employee Management System, where employees belong to various departments within a company. Spring configuration helps us writing clean and modular code.
+{{% /info %}}
+\
+Enter Spring's core concepts: `@Bean` and `@Configuration`.
+
+## Understanding @Bean
+In Spring, `@Bean` transforms ordinary methods into Spring-managed beans. These beans are objects managed by the Spring IoC container, ensuring centralized control and easier dependency management. In our Employee Management System, `@Bean` helps create beans representing employees and departments.
+
+Examples of using `@Bean`:
+
+```java
+public class Employee {
+    // Employee properties and methods
+}
+
+public class Department {
+    // Department properties and methods
+}
+
+@Configuration
+@ComponentScan(basePackages = "io.reflectoring.javaconfig")
+public class JavaConfigAppConfiguration {
+
+    @Bean(name = "newEmployee", initMethod = "init", destroyMethod = "destroy")
+    @Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+    @Cacheable(cacheNames = "employeesCache")
+    public Employee newEmployee(final String firstName, final String lastName) {
+        return Employee.builder().firstName(firstName).lastName(lastName).build();
+    }
+
+    @Bean
+    @Scope(ConfigurableBeanFactory.SCOPE_PROTOTYPE)
+    public Department newDepartment(final String deptName) {
+        final Department department = Department.builder().name(deptName).build();
+        acme().addDepartment(department);
+        return department;
+    }
+
+    @Bean(name = "founder")
+    @Qualifier(value = "founder")

```

Comment:
There's still no explanation of the `@Qualifier` annotation in the text anywhere.

Reply:
Added notes on Qualifier.

Reply:
See my comment below about `@Qualifier`.

Reply:
Also, I think I mentioned it some other comment: **the `@Qualifier` annotation has no effect on `@Bean` methods, as far as I know**! It only has an effect on _parameters_ to those methods (i.e. when we try to wire a bean, not when we define a bean). Please check this and correct all the occurrences of `@Qualifier` in the article if needed.

Reply:
Moved the explanation to `Overriding Default Bean Name`

Reply:
I removed the `@Qualifier` annotation on the `@Bean` methods now ... this annotation has no effect and is only confusing.